core/src/main/scala/quasar/destination/gbq/GBQFlow.scala

```diff
@@ -0,0 +1,314 @@
+
+import cats.ApplicativeError
+import cats.data.{ValidatedNel, NonEmptyList}
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.effect.concurrent.Ref
+import cats.implicits._
+
+import fs2.{Pipe, Stream}
+import cats.effect._
+
+import org.http4s.{
+  AuthScheme,
+  Credentials,
+  EntityEncoder,
+  MediaType,
+  Method,
+  Request,
+  Status,
+  Uri
+}
+import org.http4s.argonaut.jsonEncoderOf
+import org.http4s.client._
+import org.http4s.Header
+import org.http4s.headers.{Authorization, `Content-Type`, Location}
+
+import scalaz.{NonEmptyList => ZNEList}
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+import com.google.auth.oauth2.AccessToken
+
+final class GBQFlow[F[_]: Concurrent](
+    name: String,
+    schema: NonEmptyList[GBQSchema],
+    idColumn: Option[Column[_]],
+    config: GBQConfig,
+    client: Client[F],
+    refMode: Ref[F, WriteMode])
+    extends Flow[F] {
+
+  private def credsF =
+    tokenF map { token =>
+      Authorization(Credentials.Token(AuthScheme.Bearer, token.getTokenValue))
+    }
+
+  private val project =
+    URLEncoder.encode(config.authCfg.projectId, StandardCharsets.UTF_8.toString)
+
+  private val uriRoot =
+    s"https://bigquery.googleapis.com/bigquery/v2/projects/${project}"
+
+  private def tokenF: F[AccessToken] =
+    GBQAccessToken.token(config.serviceAccountAuthBytes)
+
+  def ingest: Pipe[F, Byte, Unit] = { (bytes: Stream[F, Byte]) =>
+    for {
+      eitherloc <- Stream.eval {
+        for {
+          mode <- refMode.get
+          jobConfig = formGBQJobConfig(schema, config, name, mode)
+          _ <- mkDataset.whenA(mode === WriteMode.Replace)
+          eloc <- mkGbqJob(jobConfig)
+        } yield eloc
+      }
+      _ <- eitherloc match {
+        case Right(locationUri) => upload(bytes, locationUri)
+        case Left(e) => Stream.raiseError[F](new RuntimeException(s"No Location URL returned from job: $e"))
+      }
+      _ <- Stream.eval(refMode.set(WriteMode.Append))
+    } yield ()
+  }
+
+  def delete(ids: IdBatch): Stream[F, Unit] = idColumn traverse_ { (col: Column[_]) =>
+    val strs: Array[String] = ids match {
+      case IdBatch.Strings(values, _) => values.map(x => "\"" + x + "\"")
+      case IdBatch.Longs(values, _) => values.map(_.toString)
+      case IdBatch.Doubles(values, _) => values.map(_.toString)
+      case IdBatch.BigDecimals(values, _) => values.map(_.toString)
+    }
+
+    def mkIn(strs: Array[String]): String = {
+      if (strs.isEmpty) ""
+      else s"${col.name} IN (${strs.mkString(", ")})"
+    }
+
+    def group(ix: Int, accum: List[String]): String = {
+      val nextIx = ix + 64
+      val taken = strs.slice(ix, nextIx)
+      val in = mkIn(taken)
+      val nextAccum = in :: accum
+      if (nextIx > ids.size + 1)
+        nextAccum.mkString(" OR ")
+      else
+        group(nextIx, nextAccum)
+    }
+
+    val groupped = group(0, List())
+
+    val where = if (groupped.isEmpty) "" else s"WHERE $groupped"
```

Comment:
It looks like if `grouped` is empty this deletes the whole dataset. Shouldn't this delete nothing in that case?

Reply:
:+1: